### ISSUE TITLE
Question about recomputation in actor module

### ISSUE THREAD

**Issue by 0oshowero0** [volcengine/verl#51]
I'm curious about why we must need to do recomputation when using HybridEngine. In `generate_sequences()` line 367-372 [https://github.com/volcengine/verl/blob/main/verl/trainer/ppo/workers/fsdp_workers.py](fsdp_workers.py) , we have 
```python
if self._is_actor and recompute_log_prob:
            # we should always recompute old_log_probs when it is HybridEngine
            output.meta_info['micro_batch_size'] = self.config.rollout.log_prob_micro_batch_size
            output.meta_info['temperature'] = self.config.rollout.temperature
            old_log_probs = self.actor.compute_log_prob(data=output)
            output.batch['old_log_probs'] = old_log_probs
```

This `generate_sequences()` function should be run by the rollout module, who can directly provide the `old_log_probs` (see the commented lines in line 208 of vllm_rollout.py [https://github.com/volcengine/verl/blob/main/verl/trainer/ppo/rollout/vllm_rollout/vllm_rollout.py](vllm_rollout.py)). 

So why we must recompute it with extra overhead?

Besides, in my understanding, rollout module and actor module should be the same model at different steps, i.e., $M_{rollout}(t)=M_{actor}(t-1)$. It means that we need to maintains two sets of parameters for Actor model at the same time, and before we update the parameters of Actor, we should make sure the parameters-to-be-update are saved to update rollout model. But I didn't find relevant codes in `ray_trainer.py`. Do I miss something or I simply have some misunderstanding? Thanks!

**Comment by 0oshowero0** [comment]
New to RLHF, I sincerely thank the developers in advance for their attention 😁

**Comment by PeterSH6** [comment]
Hi @0oshowero0 , thanks for your questions!

Recompute `old_log_prob`:  This is mainly because many RL algorithms (e.g., PPO) are quite sensitive to logits. The logits computed by the inference engine (vLLM) might use different kernels from those used in training (FSDP and Megatron).
Therefore, we use the actor model to recompute `old_log_prob` for better logit alignment.

Update parameter between actor and rollout: Your understanding is correct. The implementation of syncing weight (i.e., model weight resharding) is implemented in [hybrid_engine](https://github.com/volcengine/verl/tree/main/verl/trainer/ppo/hybrid_engine) and be utilized as a Context Manager in fsdp_workers.py](https://github.com/volcengine/verl/tree/main/verl/trainer/ppo/hybrid_engine) (similar in megatron_workers.py)

**Comment by 0oshowero0** [comment]
I see! This is a very valuable practical experience. Thank you for sharing!
